File: branches/java/rtspproxy/rtsp/RtspTransport.java
/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   Copyright (C) 2005 - Matteo Merli - matteo.merli@gmail.com            *
 *                                                                         *
 ***************************************************************************/

/*
 * $Id$
 * 
 * $URL$
 * 
 */

package rtspproxy.rtsp;

/**
 * Parse the RTSP Transport header field.
 */
public class RtspTransport
{

	enum TransportProtocol {
		None, RTP, RDT
	}
	enum Profile {
		None, AVP
	}
	enum LowerTransport {
		None, TCP, UDP
	}
	enum DeliveryType {
		None, unicast, multicast
	}

	TransportProtocol transportProtocol;
	Profile profile;
	LowerTransport lowerTransport;
	DeliveryType deliveryType;

	String destination;
	String interleaved;
	int layers;
	boolean append;
	int ttl;
	short[] port = new short[2];
	short[] client_port = new short[2];
	short[] server_port = new short[2];
	String ssrc;
	String mode;

	/**
	 * Constructor. Creates a RtspTransport object from a transport header
	 * string.
	 */
	public RtspTransport( String transport )
	{
		transportProtocol = TransportProtocol.None;
		profile = Profile.None;
		lowerTransport = LowerTransport.None;
		deliveryType = DeliveryType.None;
		destination = null;
		interleaved = null;
		layers = 0;
		append = false;
		ttl = 0;
		port[0] = 0;
		port[1] = 0;
		client_port[0] = 0;
		client_port[1] = 0;
		server_port[0] = 0;
		server_port[1] = 0;
		ssrc = null;
		mode = null;

		parseTransport( transport );
	}

	private void parseTransport( String transport )
	{
		for ( String tok : transport.split( ";" ) ) {

			// First check for the transport protocol
			if ( tok.startsWith( "RTP" ) || tok.startsWith( "RDT" ) ) {
				String[] tpl = tok.split( "/" );
				transportProtocol = TransportProtocol.valueOf( tpl[0] );
				if ( tpl.length > 1 )
					profile = Profile.valueOf( tpl[1] );
				if ( tpl.length > 2 )
					lowerTransport = LowerTransport.valueOf( tpl[2] );
				continue;
			}

			if ( tok.toLowerCase() == "unicast" )
				deliveryType = DeliveryType.unicast;
			else if ( tok.toLowerCase() == "multicast" )
				deliveryType = DeliveryType.multicast;
			else if ( tok.startsWith( "destination" ) )
				setDestination( _getStrValue( tok ) );
			else if ( tok.startsWith( "interleaved" ) )
				setInterleaved( _getStrValue( tok ) );
			else if ( tok.startsWith( "append" ) )
				setAppend( true );
			else if ( tok.startsWith( "layers" ) )
				setLayers( Integer.valueOf( _getStrValue( tok ) ) );
			else if ( tok.startsWith( "ttl" ) )
				setTTL( Integer.valueOf( _getStrValue( tok ) ) );
			else if ( tok.startsWith( "port" ) )
				setPort( _getPairValue( tok ) );
			else if ( tok.startsWith( "client_port" ) )
				setClientPort( _getPairValue( tok ) );
			else if ( tok.startsWith( "server_port" ) )
				setServerPort( _getPairValue( tok ) );
			else if ( tok.startsWith( "ssrc" ) )
				setSSRC( _getStrValue( tok ) );
			else if ( tok.startsWith( "mode" ) )
				setMode( _getStrValue( tok ) );
		}
	}

	public String toString()
	{
		String s = "";
		s += transportProtocol;
		if ( profile != Profile.None ) {
			s += "/" + profile;
			if ( lowerTransport != LowerTransport.None )
				s += "/" + lowerTransport;
		}
		if ( deliveryType != DeliveryType.None )
			s += ";" + deliveryType;
		if ( destination != null )
			s += ";destination=" + destination;
		if ( interleaved != null )
			s += ";interleaved=" + interleaved;
		if ( append )
			s += ";append";
		if ( layers > 0 )
			s += ";layers=" + layers;
		if ( ttl > 0 )
			s += ";ttl=" + ttl;
		if ( port[0] > 0 )
			s += ";port=" + port[0] + "-" + port[1];
		if ( client_port[0] > 0 )
			s += ";client_port=" + client_port[0] + "-" + client_port[1];
		if ( server_port[0] > 0 )
			s += ";server_port=" + server_port[0] + "-" + server_port[1];
		if ( ssrc != null )
			s += ";ssrc=" + ssrc;
		if ( mode != null )
			s += ";mode=" + mode;
		return s;
	}

	/**
	 * @return Returns the append.
	 */
	public boolean isAppend()
	{
		return append;
	}

	/**
	 * @param append
	 *        The append to set.
	 */
	public void setAppend( boolean append )
	{
		this.append = append;
	}

	/**
	 * @return Returns the client_port.
	 */
	public short[] getClientPort()
	{
		return client_port;
	}

	/**
	 * @param client_port
	 *        The client_port to set.
	 */
	public void setClientPort( short[] client_port )
	{
		this.client_port = client_port;
	}

	/**
	 * @return Returns the deliveryType.
	 */
	public DeliveryType getDeliveryType()
	{
		return deliveryType;
	}

	/**
	 * @param deliveryType
	 *        The deliveryType to set.
	 */
	public void setDeliveryType( DeliveryType deliveryType )
	{
		this.deliveryType = deliveryType;
	}

	/**
	 * @return Returns the destination.
	 */
	public String getDestination()
	{
		return destination;
	}

	/**
	 * @param destination
	 *        The destination to set.
	 */
	public void setDestination( String destination )
	{
		this.destination = destination;
	}

	/**
	 * @return Returns the interleaved.
	 */
	public String getInterleaved()
	{
		return interleaved;
	}

	/**
	 * @param interleaved
	 *        The interleaved to set.
	 */
	public void setInterleaved( String interleaved )
	{
		this.interleaved = interleaved;
	}

	/**
	 * @return Returns the layers.
	 */
	public int getLayers()
	{
		return layers;
	}

	/**
	 * @param layers
	 *        The layers to set.
	 */
	public void setLayers( int layers )
	{
		this.layers = layers;
	}

	/**
	 * @return Returns the lowerTransport.
	 */
	public LowerTransport getLowerTransport()
	{
		return lowerTransport;
	}

	/**
	 * @param lowerTransport
	 *        The lowerTransport to set.
	 */
	public void setLowerTransport( LowerTransport lowerTransport )
	{
		this.lowerTransport = lowerTransport;
	}

	/**
	 * @return Returns the mode.
	 */
	public String getMode()
	{
		return mode;
	}

	/**
	 * @param mode
	 *        The mode to set.
	 */
	public void setMode( String mode )
	{
		this.mode = mode;
	}

	/**
	 * @return Returns the port.
	 */
	public short[] getPort()
	{
		return port;
	}

	/**
	 * @param port
	 *        The port to set.
	 */
	public void setPort( short[] port )
	{
		this.port = port;
	}

	/**
	 * @return Returns the profile.
	 */
	public Profile getProfile()
	{
		return profile;
	}

	/**
	 * @param profile
	 *        The profile to set.
	 */
	public void setProfile( Profile profile )
	{
		this.profile = profile;
	}

	/**
	 * @return Returns the server_port.
	 */
	public short[] getServerPort()
	{
		return server_port;
	}

	/**
	 * @param server_port
	 *        The server_port to set.
	 */
	public void setServerPort( short[] server_port )
	{
		this.server_port = server_port;
	}

	/**
	 * @return Returns the ssrc.
	 */
	public String getSSRC()
	{
		return ssrc;
	}

	/**
	 * @param ssrc
	 *        The ssrc to set.
	 */
	public void setSSRC( String ssrc )
	{
		this.ssrc = ssrc;
	}

	/**
	 * @return Returns the transportProtocol.
	 */
	public TransportProtocol getTransportProtocol()
	{
		return transportProtocol;
	}

	/**
	 * @param transportProtocol
	 *        The transportProtocol to set.
	 */
	public void setTransportProtocol( TransportProtocol transportProtocol )
	{
		this.transportProtocol = transportProtocol;
	}

	/**
	 * @return Returns the ttl.
	 */
	public int getTTL()
	{
		return ttl;
	}

	/**
	 * @param ttl
	 *        The ttl to set.
	 */
	public void setTTL( int ttl )
	{
		this.ttl = ttl;
	}

	/**
	 * Get the value part in a string like:
	 * 
	 * <pre>
	 * key=value
	 * </pre>
	 * 
	 * @param str
	 *        the content string
	 * @return a String containing only the value
	 */
	private static String _getStrValue( String str )
	{
		String[] list = str.split( "=" );
		if ( list.length != 2 )
			return null;

		return list[1];
	}

	/**
	 * Get the value part in a string like:
	 * 
	 * <pre>
	 * key=6344-6345
	 * </pre>
	 * 
	 * @param str
	 *        the content string
	 * @return a short[] containing only the value
	 */
	private static short[] _getPairValue( String str )
	{
		short[] pair = { 0, 0 };
		String[] list = str.split( "=" );
		if ( list.length != 2 )
			return pair;

		try {
			pair[0] = Integer.valueOf( list[1].split( "-" )[0] ).shortValue();
			pair[1] = Integer.valueOf( list[1].split( "-" )[1] ).shortValue();

		} catch ( Exception e ) {
			return pair;
		}
		return pair;
	}
}

File: branches/qt-cpp/src/rtsp/RtspTransport.cpp
/**************************************************************************
 *   Copyright (C) 2005 Matteo Merli <matteo.merli@gmail.com>
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 *   $Id$
 * 
 *   $URL$
 * 
 *****************************************************************************/
 
#include "RtspTransport.h"
 
/**
	\code
   Transport           =    "Transport" ":"
                            1\#transport-spec
   transport-spec      =    transport-protocol/profile[/lower-transport]
                            *parameter
   transport-protocol  =    "RTP"
   profile             =    "AVP"
   lower-transport     =    "TCP" | "UDP"
   parameter           =    ( "unicast" | "multicast" )
                       |    ";" "destination" [ "=" address ]
                       |    ";" "interleaved" "=" channel [ "-" channel ]
                       |    ";" "append"
                       |    ";" "ttl" "=" ttl
                       |    ";" "layers" "=" 1*DIGIT
                       |    ";" "port" "=" port [ "-" port ]
                       |    ";" "client_port" "=" port [ "-" port ]
                       |    ";" "server_port" "=" port [ "-" port ]
                       |    ";" "ssrc" "=" ssrc
                       |    ";" "mode" = <"> 1\#mode <">
   ttl                 =    1*3(DIGIT)
   port                =    1*5(DIGIT)
   ssrc                =    8*8(HEX)
   channel             =    1*3(DIGIT)
   address             =    host
   mode                =    <"> *Method <"> | Method


   Example:
     Transport: RTP/AVP;multicast;ttl=127;mode="PLAY",
                RTP/AVP;unicast;client_port=3456-3457;mode="PLAY"
 */
 
#include <QtCore>
 
RtspTransports::RtspTransports( const QString& str )
{
	QStringList list = str.split( "," );
 	QStringListIterator it( list );
 	while ( it.hasNext() ) {
 		transports << *( new RtspTransport( it.next() ) );
 	}
}
 
RtspTransports::~RtspTransports()
{
}

QString RtspTransports::toString() const
{
	QStringList list;
	RtspTransportListIterator it( transports );
	while ( it.hasNext() ) {
		list << it.next().toString();
	}
	return list.join(",");
}

/////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////////////////////////////////////////////////////////////


RtspTransport::TransportProtocol RtspTransport::getTransportProtocol( const QString& str )
{
	if ( str == "RTP" ) return RTP;
	if ( str == "RDT" ) return RDT;
	
	return TransportProtocolNone;	
}

QString RtspTransport::getTransportProtocol( TransportProtocol p )
{
	switch ( p ) {
		case RTP : return "RTP";
		case RDT : return "RDT";
		default: return "";
	};	
}

RtspTransport::Profile RtspTransport::getProfile( const QString& str )
{
	if ( str == "AVP" ) return AVP;
	
	return ProfileNone;
}

QString RtspTransport::getProfile( Profile p )
{
	switch ( p ) {
		case AVP : return "AVP";
		default: return "";
	};
}

RtspTransport::LowerTransport RtspTransport::getLowerTransport( const QString& str )
{
	if ( str == "TCP" ) return TCP;
	if ( str == "UDP" ) return UDP;
	return LowerTransportNone;
}

QString RtspTransport::getLowerTransport( LowerTransport p )
{
	switch ( p ) {
		case TCP : return "TCP";
		case UDP : return "UDP";
		default  : return "";
	};
}

RtspTransport::DeliveryType RtspTransport::getDeliveryType( const QString& str )
{
	if ( str == "unicast" ) return UNICAST;
	if ( str == "multicast" ) return MULTICAST;
	return DeliveryTypeNone;
}

QString RtspTransport::getDeliveryType( DeliveryType p )
{
	switch ( p ) {
		case UNICAST : return "unicast";
		case MULTICAST : return "multicast";
		default: return "";
	};
}


QString _getStrValue( const QString& str )
{
	QStringList list = str.split( "=" );
	if ( list.count() != 2 )
		return QString::null;
		
	return list.at( 1 );
}

Int16Pair _getPairValue( const QString& str )
{
	Int16Pair pair;
	pair.first = 0; pair.second = 0;
	QStringList list = str.split( "=" );
	if ( list.count() != 2 )
		return pair;
	pair.first = list.at( 1 ).section( "-", 0, 0 ).toInt();
	pair.second = list.at( 1 ).section( "-", 1, 1 ).toInt();
	return pair;
}

/////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////////////////////////////////////////////////////////////


RtspTransport::RtspTransport( const QString& str )
{
	qDebug() << "New Transport:" << str;
	transport_protocol = TransportProtocolNone;
	profile = ProfileNone;
	lower_transport = LowerTransportNone;
	delivery_type = DeliveryTypeNone;
	destination = QString::null;
	interleaved = QString::null;
	layers = 0;
	append = false;
	ttl = 0;
	port.first = 0; port.second = 0;
	client_port.first = 0; client_port.second = 0;
	server_port.first = 0; server_port.second = 0;
	ssrc = QString::null;
	mode = QString::null;
	
	parseTransport( str );
}

RtspTransport::~RtspTransport()
{
}
 
void RtspTransport::parseTransport( const QString& transport )
{
	QStringList tokens = transport.split( ";" );
	QStringListIterator it( tokens );
	while ( it.hasNext() ) {
		QString tok = it.next();
		
		// First check for the transport protocol
		if ( tok.startsWith( "RTP" ) || tok.startsWith( "RDT" ) ) {
			QStringList tpl = tok.split( "/" );
			transport_protocol = getTransportProtocol( tpl.at( 0 ) );
			if ( tpl.count() > 1 )
				profile = getProfile( tpl.at( 1 ) );
			if ( tpl.count() > 2 )
				lower_transport = getLowerTransport( tpl.at( 2 ) );
			continue;
		}
		
		if ( tok == "unicast" )
			setDeliveryType( UNICAST );
		else if ( tok == "multicast" ) 
			setDeliveryType( MULTICAST );
		else if ( tok.startsWith( "destination" ) ) 
			setDestination( _getStrValue( tok ) );
		else if ( tok.startsWith( "interleaved" ) )
			setInterleaved( _getStrValue( tok ) );
		else if ( tok.startsWith( "append" ) )
			setAppend( true );
		else if ( tok.startsWith( "layers" ) )
			setLayers( _getStrValue( tok ).toInt() );
		else if ( tok.startsWith( "ttl" ) )
			setTTL( _getStrValue( tok ).toInt() );
		else if ( tok.startsWith( "port" ) )
			setPort( _getPairValue( tok ) );
		else if ( tok.startsWith( "client_port" ) )
			setClientPort( _getPairValue( tok ) );
		else if ( tok.startsWith( "server_port" ) )
			setServerPort( _getPairValue( tok ) );
		else if ( tok.startsWith( "ssrc" ) ) 
			setSSRC( _getStrValue( tok ) );
		else if ( tok.startsWith( "mode" ) ) 
			setMode( _getStrValue( tok ) );
	}
}

QString RtspTransport::toString() const
{
	QString res;
	QTextStream out( &res );
	out << getTransportProtocol( transport_protocol );
	if ( profile != ProfileNone )
		out << "/" << getProfile( profile );
	if ( lower_transport != LowerTransportNone )
		out << "/" << getLowerTransport( lower_transport );
	if ( delivery_type != DeliveryTypeNone )
		out << ";" << getDeliveryType( delivery_type );
	if ( destination != QString::null )
		out << ";destination=" << destination;
	if ( interleaved != QString::null )
		out << ";interleaved=" << interleaved;
	if ( append )
		out << ";append";
	if ( layers )
		out << ";layers=" << layers;
	if ( ttl )
		out << ";ttl=" << ttl;
	if ( port.first )
		out << ";port=" << port.first << "-" << port.second;
	if ( client_port.first )
		out << ";client_port=" << client_port.first << "-" << client_port.second;
	if ( server_port.first )
		out << ";server_port=" << server_port.first << "-" << server_port.second;
	if ( ssrc != QString::null )
		out << ";ssrc=" << ssrc;
	if ( mode != QString::null )
		out << ";mode=" << mode;
	return res;
}


#ifdef TEST_RTSP_TRANSPORT

int main( int argc, char** argv )
{
	if ( argc < 2 )
		return;
	RtspTransports *t = new RtspTransports( argv[ 1 ] );
	
	qDebug() << "RESULT:";
	qDebug() << t->toString();
}

#endif

 